### PR TITLE
Return a better exception message when `regexp` query is used on a numeric field

### LINKED ISSUE #14782: Confusing exception when `regexp` query used on numeric field
PUT t/t/1
    {
      "num": 34
    }
    
    GET /_search
    {
      "query": {
        "regexp": {
          "num": {
            "value": "34"
          }
        }
      }
    }
    
returns:

   "error": {
      "root_cause": [
         {
            "type": "illegal_argument_exception",
            "reason": "expected '\"' at position 11"
         }
      ],

### PR DESCRIPTION
**Pull request by camilojd**
Fixes #14782

### CODE REVIEW

**Comment by nik9000** [comment]
I think this way to do it seems pretty sane. I'm not sure what the __rightest__ way to do it is but this will certainly get the job done.

File: core/src/test/java/org/elasticsearch/search/simple/SimpleSearchIT.java [hunk 1]
```diff
@@ -336,6 +337,19 @@ public void testTooLargeFromAndSizeBackwardsCompatibilityRecommendation() throws
                 .setFrom(DefaultSearchContext.Defaults.MAX_RESULT_WINDOW * 10).get(), 1);
     }
 
+    public void testQueryNumericFieldWithRegex() throws Exception {
+        createIndex("idx");
+        indexRandom(true, client().prepareIndex("idx", "type")
+                .setSource(XContentFactory.jsonBuilder().startObject().field("num", 34).endObject()));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/simple/SimpleSearchIT.java]: In this case you can use the `setSource` version that takes a field name and save a few characters. No big deal either way though.

Comment (camilojd) [inline core/src/test/java/org/elasticsearch/search/simple/SimpleSearchIT.java]: Ah, I missed that method overload. Will fix!

**Comment by nik9000** [comment]
Tip: github emails me whenever you comment on a pull request but not when you push a new commit. I'm not really sure why. Anyway, the upshot is that if you comment after you send the patch and mention that you did so rather than before it tends to summon the reviewer quickly.

**Comment by nik9000** [comment]
Ok! Looks good. Would you mind squashing the commits and amending the subject to include words like "Return a better exception message when `regexp` query is used on a numeric field" just so the log is easier to read? That'd be great!

Not everyone in Elasticsearch squashes commits on the way in. I tend to but not everyone does and I wouldn't require it of anyone except that the commit message on the first commit isn't super useful when you are looking at `git log`.

Thanks for doing this!

**Comment by camilojd** [comment]
@nik9000 will take the notification/squashing tips into account (I'm new to all of it). Seems like everything's ok now.

Thank you for your patience!! :+1: :smiley:

**Comment by nik9000** [comment]
Merged to master, cherry picked to 2.x as b852bd3c75cfa7c99d6c64151bfe0fba3b241458 and 2.2 as e42ed38aefac3f04357cd078cfbc4321eae5072a.

**Pull request merged**